tests/end_to_end/test_control_flow.py

```diff
@@ -38,3 +41,292 @@ def test_while_executes_and_scopes_correctly(execute):
     res = execute(SAMPLE_WHILE, artifacts=["result"])
     assert res.values["result"] == 6
     assert res.artifacts["result"] == SAMPLE_WHILE
+
+
+def test_if_should_slice_within_if(execute):
+    # TODO: We only slice within a block if we can execute it. Later, we need to be able to ensure we can perform slicing among unvisited blocks as well
+    CODE = """a = 10
+b = 20
+if a > 5:
+    a = 5
+    b = 6
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 5
+    assert res.values["b"] == 6
+    assert (
+        res.artifacts["a"]
+        == """a = 10
+if a > 5:
+    a = 5
+"""
+    )
+
+
+@pytest.mark.xfail(
+    reason="Variable overwritten in visited branch results in declaration outside slice being removed"
+)
+def test_if_should_not_slice_out_existing_variable_definitions_outside_block(
+    execute,
+):
+    CODE = """a = 10
+b = 20
+if a > 5:
+    a = 5
+    b = 6
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 5
+    assert res.values["b"] == 6
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+b = 20
+if a > 5:
+    b = 6
+"""
+    )
+
+
+def test_if_should_slice_within_else(execute):
+    # TODO: We only slice within a block if we can execute it. Later, we need to be able to ensure we can perform slicing among unvisited blocks as well
+    CODE = """a = 10
+b = 20
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    a = 100
+    b = 101
+"""
+    res = execute(CODE, artifacts=["a", "b"])
+    assert res.values["a"] == 100
+    assert res.values["b"] == 101
+    assert (
+        res.artifacts["a"]
+        == """a = 10
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    a = 100
+"""
+    )
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+if a <= 5:
+    a = 5
+    b = 5
+else:
+    b = 101
+"""
+    )
+
+
+def test_if_should_not_slice_whole_block_out_if_unexecuted_block_present(
+    execute,
+):
+    CODE = """a = 10
+b = 20
+if a >= 10:
+    a += 1
+else:
+    a -= 1
+"""
+    res = execute(CODE, artifacts=["b"])
+    assert (
+        res.artifacts["b"]
+        == """a = 10
+b = 20
+if a >= 10:
+    pass
+else:
+    a -= 1
+"""
+    )
+
+
+def test_if_nested_should_slice_within_inner_if(execute):
+    CODE = """a = 10
+b = 20
+if a > 0:
+    if a > 2:
+        a = 1
+        b = 2
+    else:
+        a = 3
+        b = 4
+else:
+    a = 5
+    b = 6
+"""
+    res = execute(CODE, artifacts=["a"])
+    assert res.values["a"] == 1
+    assert res.values["b"] == 2
+    assert (
+        res.artifacts["a"]
+        == """a = 10
+if a > 0:
```

Comment:
Given that we will have INCORRECT code that won't run with this new addition (e.g., we don't have `b` in this slide), we need to make sure this release is accompanied by ample documentation that teaches people the boundary conditions and how they can fix things themselves. CC @hogepodge

Reply:
Yep, should I add something for this somewhere? Note that without this PR, the blackbox  mechanism would also generate code which wont run on branches not seen during development (same as here but wont slice out anything inside the `if`)

Reply:
We can start adding a limitations section to this doc https://docs.lineapy.org/en/latest/guide/refactor_code/index.html. We should definitely create an accompanying discussion post detailing the limitation. We can also add existing limitations that are not related to this PR but exist in our approach, but if the scope is too big there, at least add the limitation in the if/else case.